### PR TITLE
feat(batcher): validate l1handler txs in validate_block

### PR DESCRIPTION
**Pull request by yair-starkware**

### CODE REVIEW

**Comment by yair-starkware** [comment]
> [!WARNING]
> <b>This pull request is not mergeable via GitHub because a downstack PR is open. Once all requirements are satisfied, merge this PR as a stack <a href="https://app.graphite.dev/github/pr/starkware-libs/sequencer/1921?utm_source=stack-comment-downstack-mergeability-warning" >on Graphite</a>.</b>
> <a href="https://graphite.dev/docs/merge-pull-requests">Learn more</a>

* **#1921** <a href="https://app.graphite.dev/github/pr/starkware-libs/sequencer/1921?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈
* **#1907** <a href="https://app.graphite.dev/github/pr/starkware-libs/sequencer/1907?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#1879** <a href="https://app.graphite.dev/github/pr/starkware-libs/sequencer/1879?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#1870** <a href="https://app.graphite.dev/github/pr/starkware-libs/sequencer/1870?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#1791** <a href="https://app.graphite.dev/github/pr/starkware-libs/sequencer/1791?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#1788** <a href="https://app.graphite.dev/github/pr/starkware-libs/sequencer/1788?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#1787** <a href="https://app.graphite.dev/github/pr/starkware-libs/sequencer/1787?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* `main`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @yair-starkware and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment">Graphite <b>Graphite</b></a>

**Review by Yael-Starkware — CHANGES_REQUESTED** [review]
> Reviewed 1 of 2 files at r1.
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/sequencer/1921)* status: 0 of 2 files reviewed, 1 unresolved discussion (waiting on @alonh5 and @yair-starkware)

___
*[`crates/batcher/src/transaction_provider.rs` line 153 at r1](https://reviewable.io/reviews/starkware-libs/sequencer/1921#-OBPPbiqAuVLpCPzCJyn:-OBPPbiqAuVLpCPzCJyo:b6oho97) ([raw file](https://github.com/starkware-libs/sequencer/blob/a3d7880e33c599cacc5cd6fb027906151a80caa5/crates/batcher/src/transaction_provider.rs#L153)):*
> ```XML
>             Transaction::L1Handler(tx) => self.l1_provider_client.validate(tx),
>             Transaction::Account(_) => true,
>         });
> ```

You are not using the value returned from all.



_Code quote:_
```XML
        buffer.iter().all(|tx| match tx {
            Transaction::L1Handler(tx) => self.l1_provider_client.validate(tx),
            Transaction::Account(_) => true,
        });
```

**Review by yair-starkware — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/sequencer/1921)* status: 1 of 2 files reviewed, 2 unresolved discussions (waiting on @alonh5 and @Yael-Starkware)

___
*[`crates/batcher/src/transaction_provider_test.rs` line 45 at r3](https://reviewable.io/reviews/starkware-libs/sequencer/1921#-OBQWXUm3fgXqWpTr6dD:-OBUJ6NOBcTj9nkmQ0F_:by6wco3) ([raw file](https://github.com/starkware-libs/sequencer/blob/f9ec7cacf4d0ca611a30d929c19b1107de363162/crates/batcher/src/transaction_provider_test.rs#L45)):*
<details><summary><i>Previously, alonh5 (Alon Haramati) wrote…</i></summary><blockquote>

Do you need this arg?
</blockquote></details>

I use it to check that we pass the expected transaction in `validate`
___
*[`crates/batcher/src/transaction_provider.rs` line 18 at r3](https://reviewable.io/reviews/starkware-libs/sequencer/1921#-OBQV-8R4FdjDw6AIH3J:-OBUItrZ9VKO2-CsW5Mp:bb7f2mf) ([raw file](https://github.com/starkware-libs/sequencer/blob/f9ec7cacf4d0ca611a30d929c19b1107de363162/crates/batcher/src/transaction_provider.rs#L18)):*
> ```XML
>     #[error(transparent)]
>     MempoolError(#[from] MempoolClientError),
>     #[error("L1Handler transaction validation failed for tx {0}.")]
> ```

done

**Review by alonh5 — CHANGES_REQUESTED** [review]
> Reviewed 1 of 1 files at r4, 1 of 1 files at r5, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/sequencer/1921)* status: all files reviewed, 2 unresolved discussions (waiting on @Yael-Starkware and @yair-starkware)

___
*[`crates/batcher/src/transaction_provider_test.rs` line 45 at r3](https://reviewable.io/reviews/starkware-libs/sequencer/1921#-OBQWXUm3fgXqWpTr6dD:-OBUQ3vOB8I37JwFDbeP:b-fwiqv4) ([raw file](https://github.com/starkware-libs/sequencer/blob/f9ec7cacf4d0ca611a30d929c19b1107de363162/crates/batcher/src/transaction_provider_test.rs#L45)):*
<details><summary><i>Previously, yair-starkware (Yair) wrote…</i></summary><blockquote>

I use it to check that we pass the expected transaction in `validate`
</blockquote></details>

But it's just a default l1handler tx. You can just create one in the function. If you want you can also create a l1_hander_tx function and use it everywhere in the file.

**Review by Yael-Starkware — CHANGES_REQUESTED** [review]
> Reviewed 1 of 1 files at r4.
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/sequencer/1921)* status: all files reviewed, 3 unresolved discussions (waiting on @yair-starkware)

___
*[`crates/batcher/src/transaction_provider_test.rs` line 166 at r5](https://reviewable.io/reviews/starkware-libs/sequencer/1921#-OBUiXTkEqRIbO2zlr_p:-OBUiXTkEqRIbO2zlr_q:boau177) ([raw file](https://github.com/starkware-libs/sequencer/blob/7a26e196c4800642c5882fd8f05348e9716ecc9a/crates/batcher/src/transaction_provider_test.rs#L166)):*
> ```XML
>         ))))
>         .await
>         .unwrap();
> ```

extract to a common function used by both tests.



_Code quote:_
```XML
    tx_sender.send(Transaction::L1Handler(L1HandlerTransaction::default())).await.unwrap();
    tx_sender
        .send(Transaction::Account(AccountTransaction::Invoke(executable_invoke_tx(
            InvokeTxArgs::default(),
        ))))
        .await
        .unwrap();
```

**Review by yair-starkware — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/sequencer/1921)* status: 0 of 2 files reviewed, 3 unresolved discussions (waiting on @alonh5 and @Yael-Starkware)

___
*[`crates/batcher/src/transaction_provider_test.rs` line 45 at r3](https://reviewable.io/reviews/starkware-libs/sequencer/1921#-OBQWXUm3fgXqWpTr6dD:-OBVoQXvC7Db3Hkagho_:b-896fix) ([raw file](https://github.com/starkware-libs/sequencer/blob/f9ec7cacf4d0ca611a30d929c19b1107de363162/crates/batcher/src/transaction_provider_test.rs#L45)):*
<details><summary><i>Previously, alonh5 (Alon Haramati) wrote…</i></summary><blockquote>

But it's just a default l1handler tx. You can just create one in the function. If you want you can also create a l1_hander_tx function and use it everywhere in the file.
</blockquote></details>

Done.
___
*[`crates/batcher/src/transaction_provider_test.rs` line 166 at r5](https://reviewable.io/reviews/starkware-libs/sequencer/1921#-OBUiXTkEqRIbO2zlr_p:-OBVoRWfEjA_xvYJnbf8:b-896fix) ([raw file](https://github.com/starkware-libs/sequencer/blob/4d7f1434faecc0aedadc14f2ba649084d9f5c86a/crates/batcher/src/transaction_provider_test.rs#L166)):*
<details><summary><i>Previously, Yael-Starkware (YaelD) wrote…</i></summary><blockquote>

extract to a common function used by both tests.
</blockquote></details>

Done.

File: crates/batcher/src/transaction_provider_test.rs [hunk 1]
```diff
@@ -77,12 +119,12 @@ async fn fill_max_l1_handler(mut mock_dependencies: MockDependencies) {
     assert!(data.iter().all(|tx| matches!(tx, Transaction::L1Handler(_))));
 
     let txs = tx_provider.get_txs(MAX_TXS_PER_FETCH).await.unwrap();
-    let data = assert_matches!(txs, NextTxs::Txs(txs) if txs.len() == 10 => txs);
-    assert!(data[..5].iter().all(|tx| matches!(tx, Transaction::L1Handler(_))));
-    assert!(data[5..].iter().all(|tx| matches!(tx, Transaction::Account(_))));
+    let data = assert_matches!(txs, NextTxs::Txs(txs) if txs.len() == MAX_TXS_PER_FETCH => txs);
+    assert!(data[..n_l1handler_left].iter().all(|tx| matches!(tx, Transaction::L1Handler(_))));
+    assert!(data[n_l1handler_left..].iter().all(|tx| matches!(tx, Transaction::Account(_))));
```
[the remark this answers is not in the document]
Comment (yair-starkware) [reply crates/batcher/src/transaction_provider_test.rs:124]: :thumbsdown:

**Review by yair-starkware — COMMENTED** [review]
> Dismissed @graphite-app[bot] from a discussion.
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/sequencer/1921)* status: 0 of 2 files reviewed, 4 unresolved discussions (waiting on @alonh5 and @Yael-Starkware)

**Review by alonh5 — CHANGES_REQUESTED** [review]
> Reviewed 1 of 1 files at r6, 1 of 1 files at r7, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/sequencer/1921)* status: all files reviewed, 4 unresolved discussions (waiting on @Yael-Starkware and @yair-starkware)

___
*[`crates/batcher/src/transaction_provider_test.rs` line 54 at r7](https://reviewable.io/reviews/starkware-libs/sequencer/1921#-OBVvg2q5CslBBvWc7TO:-OBVvg2q5CslBBvWc7TP:b-ojp8rg) ([raw file](https://github.com/starkware-libs/sequencer/blob/460c6d1630e069fbe7ed73988eb5070fea341be9/crates/batcher/src/transaction_provider_test.rs#L54)):*
> ```XML
>             .withf(move |tx_arg| tx_arg == &tx)
>             .returning(move |_| result);
>     }
> ```

Now you don't have to pass it as an arg. Just use it in the function.



_Suggestion:_
```XML
    fn expect_validate_l1handler(&mut self, result: bool) {
        self.l1_provider_client
            .expect_validate()
            .withf(move |tx_arg| tx_arg == test_l1handler_tx())
            .returning(move |_| result);
    }
```

**Comment by yair-starkware** [comment]
*[`crates/batcher/src/transaction_provider_test.rs` line 54 at r7](https://reviewable.io/reviews/starkware-libs/sequencer/1921#-OBVvg2q5CslBBvWc7TO:-OBZdUfK5_EoxqB8SKTT:b-la74z8) ([raw file](https://github.com/starkware-libs/sequencer/blob/460c6d1630e069fbe7ed73988eb5070fea341be9/crates/batcher/src/transaction_provider_test.rs#L54)):*
<details><summary><i>Previously, alonh5 (Alon Haramati) wrote…</i></summary><blockquote>

Now you don't have to pass it as an arg. Just use it in the function.
</blockquote></details>

I think it's better with an arg

**Review by alonh5 — APPROVED** [review]
:lgtm:

> Reviewed 4 of 4 files at r8, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/sequencer/1921#-:-OBZyNd_0dXZq9wyP9A6:bnfp4nl)* status: all files reviewed, 4 unresolved discussions (waiting on @Yael-Starkware and @yair-starkware)

___
*[`crates/starknet_batcher/src/transaction_provider_test.rs` line 173 at r8](https://reviewable.io/reviews/starkware-libs/sequencer/1921#-OBZy6EHBpHcTnSO0Wxl:-OBZy6EHBpHcTnSO0Wxm:undefined) ([raw file](https://github.com/starkware-libs/sequencer/blob/1ac7105f48d2a33df76b53fb39ff86e0d22f2df5/crates/starknet_batcher/src/transaction_provider_test.rs#L173)):*
> ```XML
>     mock_dependencies
>         .simulate_input_txs(vec![
>             Transaction::L1Handler(test_tx.clone()),
> ```





_Suggestion:_
```XML
test_tx
```

**Review by alonh5 — APPROVED** [review]
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/sequencer/1921)* status: all files reviewed, 3 unresolved discussions (waiting on @Yael-Starkware and @yair-starkware)

**Review by yair-starkware — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/sequencer/1921)* status: 3 of 4 files reviewed, 2 unresolved discussions (waiting on @alonh5 and @Yael-Starkware)

___
*[`crates/starknet_batcher/src/transaction_provider_test.rs` line 173 at r8](https://reviewable.io/reviews/starkware-libs/sequencer/1921#-OBZy6EHBpHcTnSO0Wxl:-OBdjXr30KucPyL9RrYj:bb7f2mf) ([raw file](https://github.com/starkware-libs/sequencer/blob/1ac7105f48d2a33df76b53fb39ff86e0d22f2df5/crates/starknet_batcher/src/transaction_provider_test.rs#L173)):*
> ```XML
>     mock_dependencies
>         .simulate_input_txs(vec![
>             Transaction::L1Handler(test_tx.clone()),
> ```

done

**Review by Yael-Starkware — APPROVED** [review]
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/sequencer/1921)* status: 3 of 4 files reviewed, all discussions resolved (waiting on @alonh5)

**Review by alonh5 — APPROVED** [review]
> Reviewed 1 of 1 files at r9, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/sequencer/1921)* status: <span>:shipit:</span> complete! all files reviewed, all discussions resolved (waiting on @yair-starkware)

**Pull request merged**